### PR TITLE
Add KotlinConf Event (and various improvements)

### PR DESCRIPTION
**Pull request by michael-r-may**
*Not for merging yet, but for review*

Added locale as an injected dependency
Switched to the Swift back end so we have iOS/Android Apps <-> Kotlin BFF <-> Swift API
Started adding the option to choose different conferences

### CODE REVIEW

File: src/main/kotlin/com/romainpiel/catsanddogs/Main.kt [hunk 1]
```diff
@@ -23,10 +24,51 @@ fun main(args: Array<String>) {
             get("/schedule.json") {
                 val fromStr: String? = call.request.queryParameters["from"]
                 val from = fromStr?.let { OffsetDateTime.parse(it) }
+                val locale = Locale.forLanguageTag("pl")
```
Comment (romainpiel) [inline src/main/kotlin/com/romainpiel/catsanddogs/Main.kt]: Nice, I think we could pass that local as a `Accept-Language` header

File: src/main/kotlin/com/romainpiel/catsanddogs/ScheduleRepository.kt [hunk 2]
```diff
@@ -25,20 +29,42 @@ class ScheduleRepository {
                 .create()
 
         val retrofit = Retrofit.Builder()
-                .baseUrl("https://catsanddogs-kotlin-server.herokuapp.com")
+                .baseUrl("https://catsanddogs-swift-server.herokuapp.com/")
```
Comment (romainpiel) [inline src/main/kotlin/com/romainpiel/catsanddogs/ScheduleRepository.kt]: 👌

File: src/main/kotlin/com/romainpiel/catsanddogs/ScheduleRepository.kt [hunk 3]
```diff
@@ -25,20 +29,42 @@ class ScheduleRepository {
...
 
-    fun schedule(from: OffsetDateTime?): Single<String> {
+    fun schedule(from: OffsetDateTime?, locale: Locale, conference: Conference): Single<String> {
         val safeFrom = from ?: OffsetDateTime.MIN
-        return service.getConference()
+
+        when (conference) {
+                Conference.MCE4 -> return scheduleMCE4(safeFrom, locale)
+                Conference.KotlinConf -> return scheduleKotlinConf(safeFrom, locale)
+        }
```
Comment (romainpiel) [inline src/main/kotlin/com/romainpiel/catsanddogs/ScheduleRepository.kt]: tiny thing, in kotlin you can do:
```kotlin
return when (conference) {
        Conference.MCE4 -> scheduleMCE4(safeFrom, locale)
        Conference.KotlinConf -> scheduleKotlinConf(safeFrom, locale)
}
```

File: src/main/kotlin/com/romainpiel/catsanddogs/ScheduleRepository.kt [hunk 4]
```diff
@@ -25,20 +29,42 @@ class ScheduleRepository {
...
+    }
+
+    private fun scheduleMCE4(from: OffsetDateTime, locale: Locale): Single<String> {
```
Comment (romainpiel) [inline src/main/kotlin/com/romainpiel/catsanddogs/ScheduleRepository.kt]: As we have a lot of shared code between the two methods, we could also do:
```kotlin
private fun Observable<ApiConference>.toJson(from: OffsetDateTime, locale: Locale) =
        this.map { it.schedule }
              .flatMapIterable { it }
              .filter { !safeFrom.isAfter(it.datestamp) }
              .filter { !from.isAfter(it.datestamp) }
              .map {
                  val date = it.datestamp.format(ofPattern("EEEE, d MMMM y", locale)).capitalize()
                  val time = it.datestamp.format(ofPattern("HH:mm", locale))
                  Item(it.title, it.speaker.joinToString(), date, time)
              }
              .toList()
              .toJson(gson)
```

and use it like that:
```kotlin
private fun scheduleMCE4(from: OffsetDateTime, locale: Locale) = service.getMCE4Schedule().toJson()
```

(not entirely sure the code compiles, I just wrote it here :D)

**Review by romainpiel — COMMENTED** [review]
Awesome to see you contribute 🎉

File: src/main/kotlin/com/romainpiel/catsanddogs/Main.kt (hunk 1 above)
Comment (michael-r-may) [reply src/main/kotlin/com/romainpiel/catsanddogs/Main.kt]: I've done what I think needs to be done, but it won't work. :(

File: src/main/kotlin/com/romainpiel/catsanddogs/ScheduleRepository.kt (hunk 3 above)
Comment (michael-r-may) [reply src/main/kotlin/com/romainpiel/catsanddogs/ScheduleRepository.kt]: Nice!

File: src/main/kotlin/com/romainpiel/catsanddogs/ScheduleRepository.kt (hunk 4 above)
Comment (michael-r-may) [reply src/main/kotlin/com/romainpiel/catsanddogs/ScheduleRepository.kt]: Yep

**Comment by michael-r-may** [comment]
@RomainPiel It's ready for review and deploy!

File: src/main/kotlin/com/romainpiel/catsanddogs/Main.kt [hunk 5]
```diff
@@ -25,34 +26,34 @@ fun main(args: Array<String>) {
...
+                val json = scheduleRepository.schedule(from, locale, conference).blockingGet()
                 call.respondText(json, ContentType.Application.Json)
             }
 
             get("/{conference}/schedule.json") {
-                val parameters = parameters(call.request.queryParameters)
+                val from = offsetDateTime(call.request.queryParameters)
                 val conference: Conference = Conference.instance(call.parameters["conference"]) ?: Conference.MCE4
+                val acceptLanguage: String = call.request.headers["Accept-Language"] ?: "pl-PL"
+                val locale = Locale.forLanguageTag(acceptLanguage)
```
Comment (romainpiel) [inline src/main/kotlin/com/romainpiel/catsanddogs/Main.kt]: weird that it doesn't work, I'll have a look this weekend

**Review by romainpiel — APPROVED** [review]
ship it!

File: src/main/kotlin/com/romainpiel/catsanddogs/Main.kt (hunk 1 above)
Comment (romainpiel) [reply src/main/kotlin/com/romainpiel/catsanddogs/Main.kt]: I'm curious why it won't work, I've just tested and it works for me

Comment (romainpiel) [reply src/main/kotlin/com/romainpiel/catsanddogs/Main.kt]: Oh retrofit 🤔 let me see...

Comment (romainpiel) [reply src/main/kotlin/com/romainpiel/catsanddogs/Main.kt]: So I have just tested and it looks fine to me:
```
GET /schedule.json
Accept-Language: en-UK

[
    {
        "title": "We are counting on you",
        "subtitle": "Jeff Watkins",
        "date": "Monday, 8 May 2017",
        "time": "10:00"
    },
```

```
GET /schedule.json
Accept-Language: fr-FR

[
    {
        "title": "We are counting on you",
        "subtitle": "Jeff Watkins",
        "date": "Lundi, 8 mai 2017",
        "time": "10:00"
    },
```

**Comment by romainpiel** [comment]
@michael-r-may everything looks fine to me, merge whenever you want, autodeploy is turned on on heroku

File: src/main/kotlin/com/romainpiel/catsanddogs/Main.kt (hunk 5 above)
Comment (romainpiel) [reply src/main/kotlin/com/romainpiel/catsanddogs/Main.kt]: See https://github.com/RomainPiel/CatsAndDogs-kotlin-bff/pull/3#discussion_r136696241

**Comment by romainpiel** [comment]
Merging so I can update the clients :)

**Pull request merged**